### PR TITLE
PERF-#4892: Compute `lengths` in `rebalance_partitions` when possible

### LINKED ISSUE #4892: PERF: compute `lengths` in `rebalance_partitions` when possible

### PR DESCRIPTION
**Pull request by anmyachev**
Signed-off-by: Myachev <anatoly.myachev@intel.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #4892 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date 
- [x] added (Issue Number: PR title (PR Number)) and github username to release notes for next major release

### CODE REVIEW

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 1]
```diff
@@ -1431,6 +1445,11 @@ def rebalance_partitions(cls, partitions):
                 new_partitions.append(
                     cls.column_partitions(partitions[start : stop + 1], full_axis=False)
                 )
+                length = sum(part.length() for part in partitions[start : stop + 1, 0])
+                for col in new_partitions[-1]:
+                    col._length_cache = length
```
Comment (mvashishtha) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:1450]: This seems like it's repeating code like this: https://github.com/modin-project/modin/blob/c67324b04b01e7d1f27a7c91565c488537891de1/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/virtual_partition.py#L376-L378

Do we need to have both?

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 2]
```diff
@@ -1417,10 +1423,18 @@ def rebalance_partitions(cls, partitions):
                         ],
                         0,
                     )
+                    for obj in partitions[stop + 1]:
+                        obj._length_cache = partition_size - (
```
Comment (mvashishtha) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:1427]: Does [try_recompute_cache](https://github.com/modin-project/modin/blob/c67324b04b01e7d1f27a7c91565c488537891de1/modin/core/dataframe/pandas/partitioning/partition.py#L205) in the `mask` not cover us?

Comment (anmyachev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:1427]: Good question!

This should help in the case of a regular partition, but not in the case of a virtual one, where a materialization is performed before the mask is executed, which overwrites the length/width with remote references. However, it is possible to preserve caches when performing `force_materialization`, I suppose. What do you think about?

Comment (mvashishtha) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:1427]: Ah, I see. Let's leave the task of preserving caches through virtual partition `mask` for another PR.

File: modin/core/dataframe/pandas/partitioning/partition_manager.py (hunk 1 above)
Comment (anmyachev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:1450]: Good catch!

File: modin/core/dataframe/pandas/partitioning/partition_manager.py (hunk 2 above)
Comment (anmyachev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:1428]: The issue for that: https://github.com/modin-project/modin/issues/4903

**Review by YarShev — DISMISSED** [review]

**Review by mvashishtha — DISMISSED** [review]

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 3]
```diff
@@ -1369,18 +1373,21 @@ def rebalance_partitions(cls, partitions):
...
-                            partitions[i : i + chunk_size],
-                            full_axis=False,
-                        )
-                        for i in range(
-                            0,
-                            num_existing_partitions,
-                            chunk_size,
-                        )
-                    ]
+                return (
```
Comment (jbrockmendel) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:1376]: nitpick: cleaner do do

```
new_partitions = np.array(...)
return new_partitions, None
```

**Review by YarShev — APPROVED** [review]

**Pull request merged**